cl-postgres/scram.lisp

```diff
@@ -0,0 +1,296 @@
+          (loop for x across arry counting x into y do
+               (setf (aref oct-array (+ y initial-index)) x))
+          oct-array))))
+
+(defun gen-client-nonce (&optional (nonce-length 32))
+  "Generate a random alphanumeric nonce with a default length of 32."
+  (let* ((chars "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
+         (chars-length 62)
+         (client-nonce (make-string nonce-length))
+         (crypto:*prng* (crypto:make-prng :os :seed :random)))
+    (dotimes (i nonce-length)
+      (setf (aref client-nonce i)
+            (aref chars
+                  (crypto:strong-random chars-length))))
+    client-nonce))
+
+(defun gen-client-initial-response (user-name client-nonce)
+  (when (not user-name) (setf user-name ""))
+  (format nil "n,,n=~a,r=~a" (saslprep-normalize user-name) client-nonce))
+
+(defun gen-salted-password (password server-salt iterations &key (digest :sha256) (salt-type :byte-array))
+  "Takes an password (must be an ascii string) and server salt (by default presumed byte-array but can be set for :string or :hex) and an integer iterations. Digest is presumed to be :sha256 but can be set to other valid ironclad digests. returns a byte-array"
+  (cond ((eq salt-type :string) (setf server-salt (ironclad:ascii-string-to-byte-array server-salt)) )
+        ((eq salt-type :byte-array) t)
+        ((eq salt-type :base64-string) (setf server-salt (cl-base64:base64-string-to-usb8-array server-salt)))
+        ((eq salt-type :hex) (setf server-salt (ironclad:hex-string-to-byte-array server-salt)))
+        (t (cerror "Please enter valid salt-type" "unknown salt-type in gen-salted-password")))
+  (ironclad:pbkdf2-hash-password
+   (ironclad:ascii-string-to-byte-array (saslprep-normalize password))
+   :salt server-salt
+   :digest digest
+   :iterations iterations))
+
+(defun split-server-response (response)
+  "Takes an array of bytes which are encoded in base64,  It should return a list of three alists of the form:
+ ((\"r\" . \"odaUyoz0GpB5GxXLfe2Y8SVjZEosREsxzxhtXY1jiNebxJlohG8IRD1v\")
+ (\"s\" . \"HV25Sl/1VAUF7k+Ddv42dQ==\") (\"i\" . \"4096\") where \"r\" is the server nonce,
+ \"s\" is a base64 encoded salt and \"i\" is the number of iterations for the hash digest.
+
+ We do not use split-sequence building the cons cell because the equal sign can appear in the nonce or salt itself."
+  (loop :for x :in (split-sequence:split-sequence #\, (cl-postgres-trivial-utf-8:utf-8-bytes-to-string response))
+     :collect (let ((split-on (position #\= x)))
+                (cons (subseq x 0 split-on)
+                      (subseq x (1+ split-on))))))
+
+(defun validate-server-nonce (server-nonce client-nonce)
+  "checks whether the server-nonce begins with the client-nonce. Both need to be normal strings."
+  (if (= 0 (search client-nonce server-nonce))
+      t
+      (error 'protocol-violation
+             :message "Client-nonce not found at beginning of server-nonce")))
```

Comment:
The `t` return value is not required here.

Reply:
Agreed